### ISSUE TITLE
Redundant null guards

### ISSUE THREAD

**Issue by chrisseaton** [oracle/truffleruby#2455]
After #2454, for this test program

```ruby
def foo(time)
  time.strftime('%m')
end

loop do
  foo(Time.now)
end
```

We get this load and null guard:

<img width="900" alt="Screenshot 2021-10-05 at 19 03 14" src="https://user-images.githubusercontent.com/341094/136078478-1b139566-0791-4f06-ba11-80fcf0b103f9.png">

But that load is from:

```java
@CompilationFinal(dimensions = 1) private static final LeafRope[] PADDED_NUMBERS = createPaddedNumbersTable();
```

**Comment by chrisseaton** [comment]
Note that the null check actually comes from later code:

```java
            if (rope == null) {
                rope = appendRope;
            } else {
                rope = concatNode.executeConcat(rope, appendRope, UTF8Encoding.INSTANCE);
            }
```

But why isn't the result of the load constant? Or stamped as non-null?

**Comment by chrisseaton** [comment]
I guess I figured it out - there is nothing that knows about the aggregate contents of the array - nothing to track that no entries are null. So this is a good example of what do we do here if we don't have unsafe stamp injections?

**Comment by eregon** [comment]
The index is not a PE constant here, hence the field read is still in the code and not folded.
And indeed there is nothing tracking "are all entries non-null".
Of course `@CompilationFinal` only has effect on PE and nothing else.

Maybe PE could do this when it sees a load from a `@CompilationFinal` array with a variable index.
But it would be unsafe because technically anything could write a `null` to that array (there are no read-only arrays in Java).
So I think it's unavoidable here.

We could move this issue to oracle/graal to ask Truffle/compiler people if you like?

**Comment by eregon** [comment]
I guess one possible way is PE would take a private copy of that array, and somehow embed it in the generated code (if it's not too big, 100 elements is already not so small). But the increased footprint and complexity (also on the semantic model of PE) would likely not be worth it.

Part of the confusion might be that the graph shows the array's content, but that's only the contents at the time of PE, it could change later. I think the graph even shows the array's content even if only the array instance is `final`.

**Comment by chumer** [comment]
I don't have an idea how to do this safely without requiring a wrapper for the Object[]. If that is acceptable then we can expose a primitive for that. We can leave it open for subclasses, that will probably be needed.

Note that some Truffle framework classes actually use unsafe stamp injection. But we are extremely conservative in where we use it. I will object to exposing unsafe stamp injection to languages as an API.

**Comment by chrisseaton** [comment]
I can probably manually restructure to avoid this null check in this case.

It's an example of where there may be a requirement.

What's your idea for wrappers, Christian?

**Comment by chumer** [comment]
Some class in Truffle:

```
static class NonNullArray {
  
   private final Object[] array;   

   private ArrayWrapper(Object[] array) {
      Objects.requireNonNull(o);
      for (Object o : array) {
          Objects.requireNonNull(o);
      }
      this.array = array,
   }

  public void get(int index) {
     Object v=  array[index];
     if (CompilerDirectives.inCompiledCode()) {
        v =  injectUnsafeNonNull(v); // internal truffle API
     }
     return v;
  } 

  public void set(int index, Object value) {
    if (value == null) {
         CompilerDirectives.transferToInterpreterAndInvalidate();
         throw new NullPointerException();
    ]
    array[index] = value;
  }
}
```

But I fear the cost of the additional indirection will eat the benefit of getting rid of the null check.

**Comment by eregon** [comment]
> But I fear the cost of the additional indirection 

That would only be a cost in the interpreter, in PE code it'd be the same, at least for this use case (the array is stored in a `static final` field of RopeConstants.
`NonNullArray` sounds like a good approach to me.

**Comment by chumer** [comment]
@eregon feel free to propose it.
You can access the injection with Engine|Instrument|Utilities|LanguageAccessor.RUNTIME.unsafeCast

**Comment by eregon** [comment]
One trade-off in this case though with that wrapper is we'd have a class check when reading to cast Object -> LeafRope.
Probably it's avoidable if we also pass the class to the constructor (and then internally use a LeafRope[] or use the unsafeCast to inject the class as well).

**Comment by chumer** [comment]
@eregon we could do the same by storing a Class<?> in the array wrapper and check any value that would flow in.

**Comment by chumer** [comment]
> Also the question whether the internal array should CompilationFinal/whether the interface should be read-only.

No. If your array is read-only you should use a @CompilationFinal(dimensions=1) Object[] array.

**Comment by chumer** [comment]
Btw.: This is related to what the Static Object Model can do for you too. You can create an object descriptor and any read from a field will have the property type unconditionally without check on the field read result. Only the receiver needs a type check.

**Comment by eregon** [comment]
> No. If your array is read-only you should use a @CompilationFinal(dimensions=1) Object[] array in context or language.

But then you need both this wrapper if the index is not PE constant and that `@CompilationFinal` copy if the index is PE constant (both in memory all the time), and you need to choose the proper one for each access for efficiency.
The array here is already `@CompilationFinal(dimensions=1)` but it's not enough to avoid the null check (when the index is not PE constant).

In this case the index is probably never PE constant, but in other cases it would be hard to know, and it might become PE constant based on the guest code.

I don't think the Static Object Model is usable here, it's a mapping from indices 0-99 to a 2-characters padded Strings/Ropes.
